### PR TITLE
ADR 009: Guidelines for interaction tests in primer-react

### PR DESCRIPTION
**Pull request by pksjce**
Describe your changes here.

This ADR proposes some instructions for how we should write interaction tests. Feedback welcome!

### CODE REVIEW

**Comment by lesliecdubs** [comment]
Thanks for contributing this ADR, @pksjce!

In order to help move a decision forward, I'm assigning the following:

**ADR decision maker:** @colebemis  
**Decision deadline:** EOD Monday, 22 August

Your role as decision maker is to gather feedback, probe for discussion and disagreement, and then make an informed decision based on the feedback, identified risks, and trade-offs about whether to adopt the ADR or not. Your role is not to decide based on your own personally preferred approach.

Keep in mind that ADR decisions can be reversed or changed in the future if new information is exposed that make the ADR worth revisiting.

You can fulfill your decision maker role by doing the following:
- review the ADR, comments, and feedback closely
- seek additional feedback from any necessary parties, as needed; for example, request a review from any individuals or teams that you know would be unduly affected by this ADR, or who have extensive experience in using either the preferred or non-preferred approach and may have perspective to share about it
- follow up on any remaining concerns, questions, or unexamined risks around accepting the proposed approach
- on or before the deadline, make a decision as to whether the ADR is accepted ✅ , which you should signal by approving this PR, or rejected 🚫 , which you should signal by closing this PR with a comment explaining the reason(s)

File: contributor-docs/adrs/adr-009-interaction-tests.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+## Status
+
+Draft
+
+## Context
+
+Our testing infrastructure consists of Jest "unit" tests and automatic Chromatic visual testing. Both of these are super useful but we have been experiencing a few setbacks.
+Unofficially we test more complex interactive scenarios using our storybooks setup.
+
+1. Repeated markup in storybooks and jest tests - Since there is not automation of interaction in storybooks, we tend to test them out manually in stories and then transfer to jest. This results in us constructing similar test case scenarios in storybooks and in tests.
```
Comment (siddharthkp) [inline contributor-docs/adrs/adr-009-interaction-tests.md:12]: 10/10 can confirm. Feel like a caveman doing this.

File: contributor-docs/adrs/adr-009-interaction-tests.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+2. Debugging complex jest tests - There is no visual aspect to jest tests. Mostly everything is CLI based. This makes it hard to debug broken tests. Visibility into these tests is also low for people who want to enhance them.
+
+## Decisions
+
+Going forward, we can use [storybook's play functions](https://storybook.js.org/docs/react/writing-stories/play-function) to write interactive tests for our components. Here's a document that describes h[ow to add and run these tests](../storybook-tests.md).
+Guidelines to make this easier -
+
+1. Create a new storybook file for your components and call them "interactions". Inside these files, we can import the stories from "examples" or "fixtures" and tack on play functions to them. This will create duplicate stories but with the added benefit of the tests appearing in the interactions add-on tab. This tab will let you step through your tests step by step, increasing visibility and debuggability.
```
Comment (siddharthkp) [inline contributor-docs/adrs/adr-009-interaction-tests.md:21]: For me, it was useful to see the demo in https://github.com/primer/react/pull/2172 to understand this part. [Live demo](https://primer-e83bf66799-13348165.drafts.github.io/storybook/?path=/story/forms-form-controls-autocomplete--default)

**Review by siddharthkp — APPROVED** [review]
Approved! from my side 

I like the approach a lot more than writing interactive tests that run with jest. My only fear (probably irrational?) with this approach is that by going deeper into the storybook ecosystem, we might get locked in with the features they do or don't provide?

(The other visual alternative would be to use playwright which run a chromium instance that can be intercepted to debug (memex uses this approach), but the dev experience is not as smooth as the one in @pksjce's demo)

Finally, an unsolved problem that we'll need to figure out: How do we add results from these tests to our [coverage metrics](https://github.com/primer/react/blob/main/jest.config.js#L5)?

File: contributor-docs/adrs/adr-009-interaction-tests.md (hunk 2 above)
Comment (broccolinisoup) [reply contributor-docs/adrs/adr-009-interaction-tests.md:21]: Thanks for sharing this @siddharthkp  🙌🏼

**Comment by broccolinisoup** [comment]
Thanks so much for putting this ADR together @pksjce ! It helped me understand where we are at testing and what our options are. I haven't used Storybook's interaction tests before but I like the approach. Are we thinking them as an alternative to E2E testing or are we thinking starting with Storybook play functions and down the road, we can introduce a more comprehensive tool for E2E and integrate into our CI/CD?

File: contributor-docs/adrs/adr-009-interaction-tests.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+Our testing infrastructure consists of Jest "unit" tests and automatic Chromatic visual testing. Both of these are super useful but we have been experiencing a few setbacks.
+Unofficially we test more complex interactive scenarios using our storybooks setup.
+
+1. Repeated markup in storybooks and jest tests - Since there is not automation of interaction in storybooks, we tend to test them out manually in stories and then transfer to jest. This results in us constructing similar test case scenarios in storybooks and in tests.
+
+2. Debugging complex jest tests - There is no visual aspect to jest tests. Mostly everything is CLI based. This makes it hard to debug broken tests. Visibility into these tests is also low for people who want to enhance them.
+
+## Decisions
+
+Going forward, we can use [storybook's play functions](https://storybook.js.org/docs/react/writing-stories/play-function) to write interactive tests for our components. Here's a document that describes h[ow to add and run these tests](../storybook-tests.md).
```
Comment (lesliecdubs) [inline contributor-docs/adrs/adr-009-interaction-tests.md:18]:
```suggestion
Going forward, we can use [storybook's play functions](https://storybook.js.org/docs/react/writing-stories/play-function) to write interactive tests for our components. Here's a document that describes [how to add and run these tests](../storybook-tests.md).
```

Typo :)

File: contributor-docs/adrs/adr-009-interaction-tests.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+Guidelines to make this easier -
+
+1. Create a new storybook file for your components and call them "interactions". Inside these files, we can import the stories from "examples" or "fixtures" and tack on play functions to them. This will create duplicate stories but with the added benefit of the tests appearing in the interactions add-on tab. This tab will let you step through your tests step by step, increasing visibility and debuggability.
+
+2. Port complex jest tests onto interactive tests. This is up to developer discretion. My recommendation is to have basic "unit" tests like component existence and axe tests in jest. Anything that can benefit from a visual aspect to it can be moved to interaction tests.
+
+3. Interaction tests are not perfect. Some aspects like mocking can be challenging and we will find solutions to these as when the need arises.
```
Comment (lesliecdubs) [inline contributor-docs/adrs/adr-009-interaction-tests.md:25]: Out of curiosity, did we consider any alternative options to Storybook play functions? 

If so, it'd be nice to add a section about what else we considered. If not, all good.

File: contributor-docs/adrs/adr-009-interaction-tests.md (hunk 2 above)
Comment (pksjce) [reply contributor-docs/adrs/adr-009-interaction-tests.md:21]: @siddharthkp - Do you think I should link this stuff in the ADR?

File: contributor-docs/adrs/adr-009-interaction-tests.md (hunk 4 above)
Comment (pksjce) [reply contributor-docs/adrs/adr-009-interaction-tests.md:25]: This is a good idea. I did have thoughts around cypress and playwright before this. Not sure how much in detail I should mention them. Will add to this draft.

**Comment by pksjce** [comment]
@siddharthkp - Thanks so much for your comments.

Re- Deep storybook integration. 
We are already pretty deep into storybooks. We use them for all primer properties, most people are familiar with them, chromatic runs on top of it etc. Also Storybook is a pretty active project with pretty responsive maintainers. So we know they are going to evolve pretty fast with the general javascript ecosystem. eg - They are working on vitejs integration which could make the storybook build itself super fast. 
There's been no pushback or nobody has really proposed an alternative to storybook in our team or in the general community so far IMO. Most component libs use storybooks too. 
So, I kinda feel like its in our benefit to grow with them and make use of all their new features eg play functions. You can see that they are really working on it and some exciting updates are coming in 7.0.0 too. 

Re- Test Coverage - I've been looking into this since the last time you mentioned it in chat. I'm having a slightly hard time coming up with a solution for this. Maybe we could pair sometime next week and brainstorm what to do for this?

**Comment by siddharthkp** [comment]
> Re- Test Coverage - I've been looking into this since the last time you mentioned it in chat. I'm having a slightly hard time coming up with a solution for this. Maybe we could pair sometime next week and brainstorm what to do for this?

Yep, I'm pretty clueless as well, let's do that :)

**Pull request merged**